### ISSUE TITLE
Allow mutable function parameters with the mut keyword.

### ISSUE THREAD

**Issue by wrightwriter** [gpuweb/gpuweb#4113]
Related: https://github.com/gpuweb/gpuweb/issues/1171
This rule has a purpose, but it means porting code between other shading languages and WGSL is complicated and many people prefer being able to have mutable args.
Right now, you have to create a separate mutable variable inside the function:
```rust
fn f(a: f32) -> f32 {
  var b  = a;  
  b += 1;
  return b;
}
```
That doesn't read well and I think rust's solution to this is quite elegant - have the mut keyword allow mutability.
```rust
fn f(mut a: f32) -> f32 {
  a += 1;
  return a;
}
```

**Comment by wrightwriter** [comment]
My proposal is unrelated to inout and pointers. The equivalent in glsl would be:
```glsl
float f(float a){
  a += 1;
  return a;
}
```

**Comment by litherum** [comment]
Oh, whoops, I misunderstood. You’re right, that’s a totally different thing. Sorry!!!

**Comment by wrightwriter** [comment]
By the way, this is used in almost every shader on shadertoy, especially pretty much every raymarcher and constantly in SDFs or intersectors. I think it would add convenience and also improves on GLSL by being immutable by default. It's very easy to see which variable will be changed once you spot the `mut` keyword.

**Comment by ben-clayton** [comment]
I'm not strongly opinionated on whether this should be accepted or not, but if it is, `var` is a far better name than introducing another keyword to the language.

**Comment by litherum** [comment]
Yeah. Looking back, I'm actually kind of surprised we picked immutability to be the default for function parameters, given that every popular human-writing shading language does the opposite.

**Comment by jimblandy** [comment]
I wasn't there, but I'm going to guess two sources of precedent:

- SPIR-V parameters are immutable.
- Rust adopted immutability by default, and while it doesn't contribute to the safety model (you can just declare any variable `mut` if you want), it is something people like.

**Comment by litherum** [comment]
We discussed this in today's WGSL standardization group call. We came away with 2 resolutions:

1. This may be a good idea, but we don't think it's high enough priority to pursue _now_. We'll reconsider during v3 time.
2. What about if we added a way to shadow function parameter names? So you could do:
```
fn f(a: f32) -> f32 {
  var a  = a; // something like this
  a += 1;
  return a;
}
```
Would that be sufficient?

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2023-05-30-minutes#allow-mutable-function-parameters-with-the-mut-keyword--issue-4113">WGSL 2023-05-30 Minutes</a></summary>

* KG
* DN: I personally think it’s fine, but internally we have different opinions.
* DN: It’s tempting to say “just do `var foo = foo`” but we have shadowing rules that forbid this.
* DN: ?? said “don’t use mut use var” (JB: +1)
* DN: Also can’t do pointers, because they aren’t mutable anyways.
* JB: Fine with anything here. `mut` is fine, `var` is fine, also fine with permitting local shadowing of params. First preference is leave as is
* MM: Confused why we don’t want to do anything. I’m not against inaction, but not sure why we’re choosing it. 
* KG: Someone coming to us with a desire, even if implementable, doesn’t compel us to implement it. There’s also shading. Strong preference to do nothing now, and maybe revisit later.  Also could be a preference to resolve on “you already have a way to do this, especially if we do allow you to locally shadow params.”
* MM: Do other languages allow you to shadow params?
* KG: There are often different rules between param shadowing and local to local shadowing.
* KG: I love being able to shadow one thing with another of the same name but with different attributes like const-ness.
* JB: Rust does this all the time. It’s useful and helpful. Val redeclaration The object is not a new thing, it’s the same object, and tweaking the non-value characteristics of it.
* MM:Swift has something similar.  For optional arguments.  
    * guard let foo = foo else {return}. 
    *  It’s kinda similar in making new variable that shadows the other one, but another way to look at it is to say this is just removing the optionality.
* MM: Not gonna fight for this.
* KG: **Consensus is no change now.  But have couple of open choices if we want to make this more ergonomic.  Revisit later**.
* JB: Reason I’m not hot on doing this is because of workload.  Not being an advocate for user preference.
* MM: Would like to say: **May be a good idea, but not high enough priority for right now**.
* DS: Did we mention the possibility of shadowing parameters?
* (Heads nodding no)
* KG: Would still want to punt that into the “not now” category.
* JB: Personally **would like that feature, but not right now**.

</details>

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2023%E2%80%9012%E2%80%9005-Minutes#mutable-function-parameters-with-the-mut-keyword-4113">WGSL 2023-12-05 Minutes</a></summary>

* DN: We don’t need them, we shipped without it. It’s syntactic sugar, a var by another name. Not all that urgent.
* **->M3**

</details>

**Comment by ch3coohlink** [comment]
It's so annoying to use an another name just for a mutable function parameter!
